### ISSUE TITLE
Tagged Services (Accumulate services of abstract type)

### ISSUE THREAD

**Issue by Kasheen** [gracicot/kangaru#81]
**Is your feature request related to a problem? Please describe.**
I don't know if this is possible already but I'm looking for a way to register an abstract service, then a bunch of concrete services. I then have another service which depends on ALL available concretes of the abstract service.

Essentially the main service is "pluggable" by creating strategies for it to use.

Perhaps you have an encryption service which has strategies that can do RSA, or Blowfish.
Perhaps you have a sorting service which has strategies that can do quicksort, merge sort, tree sort.
Perhaps you have a file meta scanner which has strategies that can scan tags from mp3, flac, ogg.

**Describe the solution you'd like**
The above are arbitrary examples but hopefully it explains the use case. Sometimes you create a strategy for a service to use which you don't want to have to manually wire every time the service is created, you just want them injected.

Perhaps when a service has a constructor which has a `vector<AbstractStrategyService> strategies` the injector could inject all such strategies which have been mapped via `kgr::overrides<AbstractStrategyService>`.

**Describe alternatives you've considered**
At the moment it seems like I need to create a strategy factory, then I need to make a factory of factories service, then manually wire those strategy factories with the main factory which gets injected into the real service which uses the strategies.

Perhaps I'm thinking about this the wrong way or I'm missing a feature that allows this by default.

**Additional context**
The reason I've called this "tagged services" is because symfony 2 webframework has a DIC where you can "tag" a service as a certain type and have all those services with a given tag injected into another service.

**Comment by gracicot** [comment]
Hello, thanks for reporting! Before anything I wonder, is it related to [this issue](https://github.com/gracicot/kangaru/issues/72)? Or really you wish to accumulate all services realed to something?

**Comment by Kasheen** [comment]
Hello, I took a look at #72 and I see where this sounds like a "tagged" service. (I only called this ticket that name because it's [how it works in symfony framework](https://symfony.com/doc/current/service_container/tags.html#reference-tagged-services). You can accumulate all services by a given meta tag, but that makes sense in PHP because it's dynamically typed, in C++ it'd make no sense.

I really want to accumulate all the services overriding something. I wrote some pseudo code, maybe it will demonstrate why this might be useful.

````cpp
// This class would be in a core module.
struct Tag
{
	std::string artist;
	std::string trackTitle;
}

// This class would be in a core module.
class AudioTagScannerStrategy
{
public:
	virtual ~AudioTagScannerStrategy() = default;

	virtual std::vector<Tag> scanDirectory(std::string directoryPath) const = 0;
};

// This class would be in a module handling MPEG.
class MpegAudioTagScannerStrategy : public AudioTagScannerStrategy
{
public:
	std::vector<Tag> scanDirectory(std::string directoryPath) const override
	{
		// make empty vector
		// open directoryPath
		// find files which are *.mp3
		// read their tag (mp3 [id3] specific code) and add to vector
		// return vector
	}
};

// This class Would be in a module handling FLAC.
class FlacAudioTagScannerStrategy
{
public:
	std::vector<Tag> scanDirectory(std::string directoryPath) const override
	{
		// Same as MPEG scanner but finds flac files and the actual tag on the file
		// needs to be read differently.
	}
};

// This class would be in a core module.
class AudioTagScanner // : public AudioTagScannerStrategy (see note below)
{
	vector<AudioTagScannerStrategy> mStrategies;
	
public:
	// Inject ALL registered strategies directly
	explicit AudioTagScanner(vector<AudioTagScannerStrategy> strategies)
		: mStrategies(strategies)
	{}

	// Now this bit can be a bit odd because technically the AudioTagScanner also
	// implements AudioTagScannerStrategy but you'd likely not register it as
	// overriding AudioTagScannerStrategyService, it's more top level.
	std::vector<Tag> scanDirectory(std::string directoryPath) const
	{
		// Make empty vector<Tag>

		// For each mStrategy
			// vector.append(strategy.scanDirectory(directoryPath) )
	
		// return vector;
	}
};

// Mapping each of those services to regular non-single service.

int main() {
	kgr::container container;

	// Injects all dependencies into the scanner because AudioTagScanner wants
	// a vector of strategies. Without being able to do this injection it seems
	// that any code using an AudioTagScanner service now needs to know every
	// strategy that any module of the application has supplied and inject manually.
	// Or we need another factory service which each module registers a strategy factory on
	// meaning more boilerplate.
	container.service<AudioTagScannerService>().scanDirectory("/usr/user/music"));

	return 0;
}
````

#72 Might be useful if it could accumulate all services under a given meta descriptor - but even then in my example `AudioTagScanner` would need to receive the `container` to then query the meta descriptor to receive its strategies instead of just having them injected.

Thanks for your work on this library by the way. I love that it's quite well documented and reasonably easy to get started with.

**Comment by gracicot** [comment]
You're welcome, and thank you for providing this example! I'm quite busy recently and I haven't had time to put as much effort on this library as I did before, but I'll see if a workaround could be done to reduce the impact of your case.

I plan to change the map container inside the `kgr::container` for the next version to allow that, something like a `flat_multimap`. The STL's `std::unordered_multimap` don't cover the use case I need, which is to retrieve the last inserted element of a key. The "multi" is the missing thing required for the container to know that there can be many instances available for a service.

**Comment by Kasheen** [comment]
Thanks again for the quick reply, also thanks for confirming what I wanted to do is not possible with the container, I just wanted to make sure that I wasn't missing anything.

For now I think I will go with my current work around which is to inject a factory-of-factories (`MultiAudioTagScannerStrategyFactory`) single service into the `AudioTagScanner` non-single service in my example. Each module that provides a strategy can just retrieve the factory-of-factories single service and register a sub-factory which generates a concrete strategy - code for that should be minimal since I can make those sub-factories from a template class and then just define a class like `class MpegAudioTagScannerStrategyFactory : public AudioTagScannerFactory<MpegAudioTagScannerStrategy>`.

If it's something you support in the future I think it might be a useful feature.

Thanks again for your help.

**Comment by Kasheen** [comment]
Continuing to integrate kangaru into my project I've realised that this idea might not even be feasible/useful in c++.

Lets say you did have a class that wanted to make an `AudioTagScanner` service which relied on some strategies as above. When you attempt to call `container.service<AudioTagScanner>()` the `.cpp` file making that call needs to have included `AudioTagScanner.h` and all other dependency headers (i.e. strategies supplied from other modules).

There might be a use case where the strategies are single services and the container could inject those already instantiated single services, but that may lead to odd issues because you'd need to "start" each of the strategies before making the scanner.

I notice that Boost.DI has a [feature request](https://github.com/boost-experimental/di/issues/340) very much like this one, but one of the devs thinks it isn't possible. When mapping services to an abstract type it's possible to supply a list of concrete types in that library, but once that's defined it can't be added too, only overridden like in kangaru it seems.

Hypodermic [has a feature](https://github.com/ybainier/Hypodermic/wiki/Resolving-components#resolve-all-components-of-a-kind) to inject multiple services of an abstract base class, but I think the way it does mapping is different from kangaru, it seems like maps can be assembled and passed about, I wonder if the map generates some kind of template factory class to make its services.

Just thought I'd mention this in case such a feature was ever attempted to be implemented.

**Comment by gracicot** [comment]
@Kasheen Thanks for linking other issues. You research the subject and I'm grateful for it!

I managed to find some time and work a solution that would be non-intrusive to your code. That is, you don't need to create factories and manual injection, but instead make an extension to the library.

It's not as memory efficient as if I contained the services in a multimap and it has a few rough edges, but at least it's as automatic and require just a few modification on the mapping:

```c++
template<typename T>
struct AllServiceImplementing : kgr::single_service<std::vector<typename std::remove_reference<kgr::service_type<T>>::type*>> {};

template<typename, typename = kgr::overrides<>, typename = kgr::dependency<>>
struct AccumulatedService;

template<typename T, typename... Parents, typename... Deps>
struct AccumulatedService<T, kgr::overrides<Parents...>, kgr::dependency<Deps...>> : kgr::single_service<T, kgr::dependency<AllServiceImplementing<Parents>..., Deps...>>, kgr::overrides<Parents...> {
	using parent = kgr::single_service<T, kgr::dependency<AllServiceImplementing<Parents>..., Deps...>>;
	
	AccumulatedService() = default;

	template<typename... Args>
	AccumulatedService(kgr::in_place_t, kgr::service_type<AllServiceImplementing<Parents>>... accumulators, Args&&... args) :
		parent{kgr::in_place_t{}, std::forward<Args>(args)...}
	{
		(void) std::initializer_list<int>{(accumulators.emplace_back(&this->instance()), 0)...};
	}
};
```

This works by creating a single service of a vector to abstract service, then make a [custom service](https://github.com/gracicot/kangaru/blob/a3272c94828d94c3006b8b625dbe8e512fcd49e9/doc/section09_definitions.md) that take the vector as dependency to inject itself into it.

It is implemented as a [generic service](https://github.com/gracicot/kangaru/blob/a3272c94828d94c3006b8b625dbe8e512fcd49e9/doc/section12_generic.md) to make it's use simpler and to not repeat yourself while doing the mapping.

Speaking of mapping, it will look like that for your example:

```c++
using AudioTagScannerStrategyService = kgr::abstract_service<AudioTagScannerStrategy>;

using MpegAudioTagScannerStrategyService = AccumulatedService<MpegAudioTagScannerStrategy, kgr::overrides<AudioTagScannerStrategyService>>;
using FlacAudioTagScannerStrategyService = AccumulatedService<FlacAudioTagScannerStrategy, kgr::overrides<AudioTagScannerStrategyService>>;

struct AudioTagScannerService : kgr::service<AudioTagScanner, kgr::dependency<AllServiceImplementing<AudioTagScannerStrategyService>>> {};
```

The accumulated services are aliases because making them struct would require to inherit parent constructor. I've hit a limitation of the library while trying to mitigate that.

When no constructor of the service definition type is available, the container will try calling the default constructor + `emplace(...)` on the type (eg. `MpegAudioTagScannerStrategyService s; s.emplace(...)`). But right now the emplace function cannot be extended since it's private, then friended to the container. I will fix that limitation in the next version. I'll simply make that member protected or public.

The other limitation is that autowiring is not possible with the current API. I'll try to make autowiring more accessible to custom services but it's a complicated subject.

Bonus: here's a [wandbox example of your snippet](https://wandbox.org/permlink/czNUO3BPxuumTGET)

**Comment by Kasheen** [comment]
Thank you so much for this workaround! That was really more than I expected. I'll give this a try in my project.

**Comment by Kasheen** [comment]
Just another note quickly, reading the issue tracker, I think this may actually be a duplicate of #71.

**Comment by gracicot** [comment]
This issue is now more elaborated than the other one. I'll close #71 and make this one a non-question.

**Comment by gracicot** [comment]
Quick update: I've started to tackle this item recently. I'll push some early work on the `dev-4.2.x` branch soon. With that and #72, I'll release `v4.2.0`. Thanks everyone for reporting these issues and making this library better!

**Comment by gracicot** [comment]
Fixed in branch `dev-4.2.x`. There's now a magic service named `kgr::override_range_service<Service>` that injects a range of implementers of that service. It cannot be autowired, but an autowired wrapper can be created to bind to vector of service or other containers.
